.devcontainer/devcontainer.json

```diff
@@ -0,0 +1,23 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/go
+{
+	"name": "Go",
+	
+	// Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+	"image": "mcr.microsoft.com/devcontainers/go:1.19-bullseye",
+
+	// Features to add to the dev container. More info: https://containers.dev/features.
+	"features": {
+		"ghcr.io/devcontainers/features/docker-in-docker": 2,
```

Comment:
When I tested by creating a codespace from this Pull Request I ran into an error with this feature. Can help look into it a bit later.

From the `creation.log`
```
#16 18.32 ERROR: Feature "Docker (Docker-in-Docker)" (ghcr.io/devcontainers/features/docker-in-docker) failed to install! Look at the documentation at https://github.com/devcontainers/features/tree/main/src/docker-in-docker for help troubleshooting this error.
------
ERROR: failed to solve: executor failed running [/bin/sh -c cd /tmp/build-features/docker-in-docker_1 && chmod +x ./devcontainer-features-install.sh && ./devcontainer-features-install.sh]: exit code: 1
2023-03-02 16:29:33.409Z: {"outcome":"error","message":"Command failed: docker buildx build --load --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.28.0-1677774539956 --build-arg _DEV_CONTAINERS_BASE_IMAGE=mcr.microsoft.com/devcontainers/go:1.19-bullseye --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp --target dev_containers_target_stage -t vsc-etcd-bd0b5d706b12267982e9db15cec4c4eb-features -f /tmp/devcontainercli-root/container-features/0.28.0-1677774539956/Dockerfile.extended /var/lib/docker/codespacemount/.persistedshare/empty-folder","description":"An error occurred setting up the container."}
2023-03-02 16:29:33.417Z: devcontainer process exited with exit code 1
2023-03-02 16:29:33.421Z: Stop: Run: docker buildx build --load --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.28.0-1677774539956 --build-arg _DEV_CONTAINERS_BASE_IMAGE=mcr.microsoft.com/devcontainers/go:1.19-bullseye --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp --target dev_containers_target_stage -t vsc-etcd-bd0b5d706b12267982e9db15cec4c4eb-features -f /tmp/devcontainercli-root/container-features/0.28.0-1677774539956/Dockerfile.extended /var/lib/docker/codespacemount/.persistedshare/empty-folder
2023-03-02 16:29:33.425Z: Error: Command failed: docker buildx build --load --build-context dev_containers_feature_content_source=/tmp/devcontainercli-root/container-features/0.28.0-1677774539956 --build-arg _DEV_CONTAINERS_BASE_IMAGE=mcr.microsoft.com/devcontainers/go:1.19-bullseye --build-arg _DEV_CONTAINERS_IMAGE_USER=root --build-arg _DEV_CONTAINERS_FEATURE_CONTENT_SOURCE=dev_container_feature_content_temp --target dev_containers_target_stage -t vsc-etcd-bd0b5d706b12267982e9db15cec4c4eb-features -f /tmp/devcontainercli-root/container-features/0.28.0-1677774539956/Dockerfile.extended /var/lib/docker/codespacemount/.persistedshare/empty-folder
2023-03-02 16:29:33.432Z:     at fie (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1915:1355)
2023-03-02 16:29:33.438Z:     at P7 (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1915:1291)
2023-03-02 16:29:33.443Z:     at processTicksAndRejections (internal/process/task_queues.js:95:5)
2023-03-02 16:29:33.446Z:     at async Fie (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1921:2093)
2023-03-02 16:29:33.454Z:     at async $f (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1921:3239)
2023-03-02 16:29:33.462Z:     at async eoe (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:2045:17247)
2023-03-02 16:29:33.471Z:     at async Qse (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:2045:16993)
```

Reply:
The features configuration isn't quite right. Changing to the following works.

```json
    "features": {
        "ghcr.io/devcontainers/features/docker-in-docker:2": {},
        "ghcr.io/devcontainers/features/github-cli:1": {}
    }
```

The issue is that `"ghcr.io/devcontainers/features/docker-in-docker": 2` is syntactic sugar for `"ghcr.io/devcontainers/features/docker-in-docker:latest": { "version": "2" }`, which pulls the latest version of the feature OCI artifact and then invokes it with an options object that asks for a version "2" of Docker/Moby tools, but there is no version "2" available:

```
#16 12.57 (!) No full or partial Docker / Moby version match found for "2" on OS debian bullseye (amd64).
```

Switching to `"ghcr.io/devcontainers/features/docker-in-docker:2": {},` pulls the major version 2 of the `docker-in-docker` feature and invokes it with the default set of options, which installs the latest available version of Docker/Moby tools. Pinning to a major version of a feature is a good practice instead of floating to `latest` to avoid accidental breaking changes. The usage and options for the `docker-in-docker` feature can be seen at https://github.com/devcontainers/features/tree/main/src/docker-in-docker.

Reply:
Thanks @craiglpeters , @jkeech - I've fixed the version pinning for features as shown above and can confirm the devcontainer runs in a codespace.

Interesting I did not get this error when using `devcontainer` via cli locally or in visual studio code locally. Either way, sorted now, thanks! 🙏🏻

Reply:
Thanks @craiglpeters , @jkeech - I've fixed the version pinning for features as shown above and can confirm the devcontainer runs in a codespace.

Interesting I did not get this error when using `devcontainer` via cli locally or in visual studio code locally. Either way, sorted now, thanks! 🙏🏻